applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp

```diff
@@ -0,0 +1,655 @@
+
+        // Positive perturbation
+        LocalSystemMatrixType element_matrix_p_perturbed;
+        element_matrix_p_perturbed.resize(matrix_size, matrix_size, false);
+        this->PerturbElementWithBasis(1.0, basisIndex, rElement, rCurrentProcessInfo);
+        rElement.CalculateLeftHandSide(element_matrix_p_perturbed, rCurrentProcessInfo);
+
+        // Negative perturbation
+        LocalSystemMatrixType element_matrix_n_perturbed;
+        element_matrix_n_perturbed.resize(matrix_size, matrix_size, false);
+        this->PerturbElementWithBasis(-2.0, basisIndex, rElement, rCurrentProcessInfo);
+        rElement.CalculateLeftHandSide(element_matrix_n_perturbed, rCurrentProcessInfo);
+
+        // Reset perturbation
+        this->PerturbElementWithBasis(1.0, basisIndex, rElement, rCurrentProcessInfo);
+
+        // Compute LHS derivative
+        noalias(rElementMatrixDerivative) = (element_matrix_p_perturbed - element_matrix_n_perturbed) / (2.0 * mFiniteDifferenceStepSize);
+
+        this->UnlockElementNodes(rElement);
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function perturbs the element with the given basisIndex
+     * @details 
+     * @param Step The step direction and size
+     * @param basisIndex The index of the basis for perturbation
+     * @param rElement The element to perturb
+     * @param rCurrentProcessInfo The current process info instance
+     */
+    // This function perturbs the element with the vector with given eigenvector index basisIndex
+    void PerturbElementWithBasis(
+        const double Step,
+        const std::size_t basisIndex,
+        Element& rElement,
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY
+
+        // Initialize is necessary for updating the section properties of shell elements
+        rElement.InitializeNonLinearIteration(rCurrentProcessInfo);
+
+        // Get DOF list
+        TElementDofsVectorType r_element_dof_list;
+        rElement.GetDofList(r_element_dof_list, rCurrentProcessInfo);
+
+        // Get PhiElemental
+        LocalSystemVectorType phi_elemental;
+        this->GetPhiElemental(phi_elemental, basisIndex, rElement, r_element_dof_list, rCurrentProcessInfo);
+
+        // Apply perturbation
+        std::size_t nodal_dof_size = r_element_dof_list.size()/rElement.GetGeometry().size();
+        for(std::size_t i_node = 0; i_node < rElement.GetGeometry().size(); ++i_node)
+        {
+            for (std::size_t i_dof = 0; i_dof < nodal_dof_size; ++i_dof)
+            {
+                const double dof_perturbation = Step*mFiniteDifferenceStepSize*phi_elemental[i_node*nodal_dof_size+i_dof];
+                auto& rp_dof = r_element_dof_list[i_node*nodal_dof_size+i_dof];
+
+                // Some elements need solution step value while others need the current coordinate. Thus perturb all!
+                // Update solution step value
+                rp_dof->GetSolutionStepValue() += dof_perturbation;
```

Comment:
```suggestion
                rp_dof->FastGetSolutionStepValue() += dof_perturbation;
```

Reply:
Is this new or an upcoming feature? So far, DOF doesn't have that function

Reply:
ah sorry somehow thought this was `Node`